### PR TITLE
Propagate dynamic int on `__setitem__`.

### PR DESCRIPTION
**Pull request by ysiraichi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106208
* __->__ #105923

Fix: #105533

This PR propagates dynamic ints used as indices for `__setitem__`. In summary, we:

- Replace the integer type for `TensorIndex` (both the enum and the corresponding
functions)
- Accordingly modify _python_variable_indexing.cpp_

cc @ezyang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/csrc/autograd/python_variable_indexing.cpp [hunk 1]
```diff
@@ -468,12 +469,16 @@ int THPVariable_setitem(PyObject* self, PyObject* index, PyObject* py_value) {
...
   // handle simple types: integers, slices
-  if (THPUtils_checkLong(index)) {
+  if (THPUtils_checkLong(index) || torch::is_symint(index)) {
     if (is_tracing && THPVariable_Check(index)) {
       recordSelectTrace(THPVariable_Unpack(index));
     }
-    dispatch_set_item(
-        self_, {at::indexing::TensorIndex(THPUtils_unpackLong(index))}, value);
+    auto symint = torch::is_symint(index)
+        ? SymInt(static_cast<SymNode>(
```
Comment (albanD) [inline torch/csrc/autograd/python_variable_indexing.cpp:477]: There no simpler way to extract the SymInt from the PyObject ??

Comment (Skylion007) [reply torch/csrc/autograd/python_variable_indexing.cpp:477]: Yeah do we not have a working cast method implementation that does? @ezyang ?

Comment (ezyang) [reply torch/csrc/autograd/python_variable_indexing.cpp:477]: py::cast should work E.g. `return py::cast<c10::SymInt>(index);`

File: aten/src/ATen/TensorIndexing.h [hunk 2]
```diff
@@ -230,22 +231,24 @@ static inline Tensor applySlice(
...
     TORCH_CHECK_INDEX(
-        !(index == 0 && dim == 0 && self_sizes->empty()),
+        !(index_value == 0 && dim == 0 && self_sizes->empty()),
         "invalid index of a 0-dim tensor. ",
         "Use `tensor.item()` in Python or `tensor.item<T>()` in C++ to convert a 0-dim tensor to a number");
 
     auto size = (*self_sizes)[dim];
     TORCH_CHECK_INDEX(
-        size >= -index && size > index,
+        size >= -index_value && size > index_value,
```
Comment (ezyang) [inline aten/src/ATen/TensorIndexing.h:249]: can't you just put some guards in here? Seems reasonable to do

Comment (ysiraichi) [reply aten/src/ATen/TensorIndexing.h:249]: Or actually issuing it to `ShapeEnv` with `index < size`?

Comment (ezyang) [reply aten/src/ATen/TensorIndexing.h:249]: Issue it to the shape env

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
needs test

**Pull request closed without merging**